File: simulacrum/src/main/java/io/github/projectet/dmlSimulacrum/inventory/SlotSimulationChamber.java
package io.github.projectet.dmlSimulacrum.inventory;

import dev.nathanpb.dml.item.ItemDataModel;
import io.github.projectet.dmlSimulacrum.item.ItemPolymerClay;
import io.github.projectet.dmlSimulacrum.util.Constants;
import io.github.projectet.dmlSimulacrum.util.DataModelUtil;
import net.minecraft.inventory.Inventory;
import net.minecraft.item.Item;
import net.minecraft.item.ItemStack;
import net.minecraft.screen.slot.Slot;

public class SlotSimulationChamber extends Slot implements Constants {
    private final int index;

    public SlotSimulationChamber(Inventory inventory, int index, int x, int y) {
        super(inventory, index, x, y);
        this.index = index;
    }

    @Override
    public boolean canInsert(ItemStack stack) {
        Item item = stack.getItem();
        return switch (index) {
            case DATA_MODEL_SLOT -> !stack.isEmpty() && item instanceof ItemDataModel && DataModelUtil.getEntityCategory(stack) != null;
            case INPUT_SLOT -> !stack.isEmpty() && item instanceof ItemPolymerClay;
            default -> false;
        };
    }
}

File: simulacrum/src/main/kotlin/dev/nathanpb/dml/simulacrum/screen/SlotSimulationChamber.kt
package dev.nathanpb.dml.simulacrum.screen

import dev.nathanpb.dml.data.dataModel
import dev.nathanpb.dml.item.ITEM_POLYMER_CLAY
import dev.nathanpb.dml.item.ItemDataModel
import net.minecraft.inventory.Inventory
import net.minecraft.item.ItemStack
import net.minecraft.screen.slot.Slot

class SlotSimulationChamber(inventory: Inventory?, index: Int, x: Int, y: Int) : Slot(inventory, index, x, y) {

    override fun canInsert(stack: ItemStack): Boolean {
        return when(index) {
            0 -> {
                !stack.isEmpty &&
                stack.item is ItemDataModel &&
                stack.dataModel.category != null
            }
            1 -> {
                !stack.isEmpty &&
                stack.isOf(ITEM_POLYMER_CLAY)
            }
            else -> false
        }
    }

}